core/tx_pool.go

```diff
@@ -211,6 +223,185 @@ func (config *TxPoolConfig) sanitize() TxPoolConfig {
+}
+
+// ReadMessage wraps corresponding method on the websocket but is safe for concurrent calling
+func (wsConnection *wsConn) ReadMessage() (int, string, error) {
+	wsConnection.rlock.Lock()
+	defer wsConnection.rlock.Unlock()
+
+	messageType, message, err := wsConnection.conn.ReadMessage()
+	return messageType, string(message), err
+}
+
+// Close wraps corresponding method on the websocket but is safe for concurrent calling
+func (wsConnection *wsConn) CloseWS() error {
+	// The Close and WriteControl methods can be called concurrently with other methods
+	return wsConnection.conn.Close()
+}
+
+// Go routine that looks for messages from the relay
+func (pool *TxPool) readWSMessages() {
+	for {
+		if pool.wsConnection.wsConnected {
+			messageType, message, err := pool.wsConnection.ReadMessage()
+			if err != nil {
+				if messageType == -1 { // message type emitted when relay ws connection is closed
+					log.Error("WS connection with relay closed")
+					pool.wsConnection.wsConnected = false
+				} else {
+					log.Error("WS error while reading the relay message: ", err.Error(), messageType)
+				}
+			}
+			var abstractMessage relayAbstractMessage
+			if err := json.Unmarshal([]byte(message), &abstractMessage); err != nil {
+				log.Error("Error while decoding relay message: ", err.Error(), nil)
+			}
+			// If relay message is of type "success", log the success message from relay
+			if abstractMessage.Type == "success" {
+				var successMessage relaySuccessMessage
+				if err := json.Unmarshal([]byte(message), &successMessage); err != nil {
+					log.Error("Error while decoding relay success message: ", err.Error(), nil)
+				}
+				log.Info(successMessage.Data)
+			}
+			// If relay message is of type "bundle", decode the payload and add bundle
+			// The sender is responsible for signing the transaction and using the correct nonce and ensuring validity
+			if abstractMessage.Type == "bundle" {
+				var bundleMessage relayBundleMessage
+				if err := json.Unmarshal([]byte(message), &bundleMessage); err != nil {
+					log.Error("Error while decoding relay bundle message: ", err.Error(), nil)
+				}
+
+				var txs types.Transactions
+
+				for _, encodedTx := range bundleMessage.Data.EncodedTxs {
+					tx := new(types.Transaction)
+					if err := tx.UnmarshalBinary(encodedTx); err != nil {
+						log.Error("Error while decoding bundle transactions: ", err.Error(), nil)
+					}
+					txs = append(txs, tx)
+				}
+				// Finally, we add the bundle sent to the tx pool
+				pool.AddMevBundle(txs, big.NewInt(bundleMessage.Data.BlockNumber.Int64()), bundleMessage.Data.MinTimestamp, bundleMessage.Data.MaxTimestamp, bundleMessage.Data.RevertingTxHashes)
+			}
+		}
+
+	}
+}
+
+func (pool *TxPool) getUTCTimestamp() string {
+	now := time.Now()
+	seconds := now.Unix()
+	return strconv.FormatInt(seconds, 10)
+}
+
+func (pool *TxPool) getWSAuthSignature() string {
+	privateKey, err := crypto.HexToECDSA(pool.config.RelayWSSigningKey)
```

Comment:
shall we sign etherbase too since it is included in the header?

Reply:
I don't think we need to sign coinbase. It's just an implementation detail of it being a header, it just as easily could have been (and might be in the future), part of a config message sent in the WS.